GS/HW: Depth feedback loops and accurate AFAIL.

TJnotJT opened the pull request:
### Description of Changes
1. Add support for depth feedback loops either via texture barriers or copies.
2. DX12 and OpenGL do depth feedback by creating a temporary color texture to hold Z and using multiple color outputs.
3. Use the depth feedback for accurate AFAIL (enabled via INI setting).

### Follow-up for separate PRs
1. (Done) In certain cases we don't need full barriers for AFAIL. The cases are detected with flags `ate_RGBA_then_Z` and `ate_RGBA_then_Z` in `GSRendererHW::DrawPrims()` for the current AFAIL system.
4. ~~Add texture barriers to DX12 instead of copies.~~ ~~Nvm, I just found out that DX12 doesn't support this for feedback loops.~~ Implemented by https://github.com/PCSX2/pcsx2/pull/13653.

### Rationale behind Changes

1. Fixes some effects that are not accurately rendered by the current AFAIL emulation. Also reduces draw calls in some cases.
2. In the future, the depth feedback will also be useful for AA1 emulation (see discussion: https://github.com/PCSX2/pcsx2/pull/13617).

### Known issues

Edit: Confirmed to not affect all systems. Seems to be fixed on my system by the recent addition of Z floor.

Unfortunately, SW depth testing can lead to severe Z fighting in some games (seemingly with drawing decals on existing geomertry), so I've made disabled depth feedback loops by default and only enabled it with maximum blending. Depth sampling, which aleady exist in master, should be unaffected by this. A rationale for keeping this PR is that is adds some common code that is useful for AA1 and ROVs.

Example in DX12: BloodRayne_SLUS-20461_20250423095228.gs.xz

<img width="512" height="448" alt="S009430_f00002_fr-1_00e00_C_24_BloodRayne_SLUS-20461_20250423095228 gs xz_shaderz" src="https://github.com/user-attachments/assets/3d8bf212-5dec-4b02-a254-40804e6d3b42" />

### Suggested Testing Steps
Please put the following in the INI or enable accurate alpha test and maximum blending in Settings>Graphics>Rendering:

```
[EmuCore/GS]
accurate_blending_unit = 5
HWAccurateAlphaTest = 1
DepthFeedbackMode = 1 # this is the default
```
Then use one of the renderers: VK, GL, DX12, DX11.

Additionally, since this PR refactors existing code to setup the alpha test, it would be good to also check basic blend for possible regressions. The following settings are suggested:

```
[EmuCore/GS]
accurate_blending_unit = 1
HWAccurateAlphaTest = 0
DepthFeedbackMode = 1 # this is the default
```

Description of new options:
- HWAccurateAlphaTest: 0 or 1. Whether to use accuracte alpha test. May lead to higher barrier and a performance penalty.
- DepthFeedbackMode:
  - 0: Disable depth feedback.
  - 1: Automatically select depth feedback mode (depth for DX11, VK; depth-as-color for DX12, GL); .
  - 2: Use depth feedback with depth texture directly (DX11, VK, GL).
  - 3: Use depth feedback by copying depth to a color texture (DX12, GL).

Fixes https://github.com/PCSX2/pcsx2/issues/3350
Fixes https://github.com/PCSX2/pcsx2/issues/3876

### Steps for testing OpenGL depth feedback (no longer required)

1. Set the following INI options:

```
[EmuCore/GS]
# Put this at the end of [EmuCore/GS] so that it overrides other settings
accurate_blending_unit = 5
HWAccurateAlphaTest = 1
DisableFramebufferFetch = true
DepthFeedbackMode = 2
```

2. Run the attached dump with OpenGL.
[God of War II_SCUS-97481_20220828133214.gs.xz.zip](https://github.com/user-attachments/files/24845497/God.of.War.II_SCUS-97481_20220828133214.gs.xz.zip)

3. Change `DepthFeedbackMode = 2` to `DepthFeedbackMode = 3`, restart PCSX2, and rerun the same dump.

Below are examples of broken vs. correct output (see main character's face and the sword).

Broken (Intel Xe iGPU, `DepthFeedbackMode = 2`)
<img width="640" height="448" alt="00786_f00003_fr1_01d00_C_24" src="https://github.com/user-attachments/assets/623552b5-e603-4ec7-92ad-c7ae9c950af2" />

Correct (Intel Xe iGPU, `DepthFeedbackMode = 3`)
<img width="640" height="448" alt="00786_f00003_fr1_01d00_C_24" src="https://github.com/user-attachments/assets/452fca03-5b95-460a-a90a-4453c38e75f8" />

Ideally, we hope mode `DepthFeedbackMode = 2` works on most systems as `DepthFeedbackMode = 3` is a slightly slower but (hopefully) more compatible option.

4. If you could post a short reply to this thread with:

- Your GPU vendor (AMD / NVIDIA / Intel)
- Your GPU model and/or architecture (e.g. RDNA2, Ampere, Xe iGPU)
- A screenshot of both outputs (or simply whether it worked correctly)

that would be greatly appreciated. Thanks in advance to anyone who can help test this!

### Did you use AI to help find, test, or implement this issue or feature?
Yes, to reference graphics API functions/semantics.

### Comparisons

There are many dumps with significant difference, but for the most part it is not obvious that they are fixes (i.e., random pixels in foliage). However, there are some clear cases below.

"everybody_golf_shadow.gs.xz" (see loading bar at bottom)

Master VK
<img width="640" height="256" alt="02978_f00001_fr1_00a00_C_32" src="https://github.com/user-attachments/assets/ed65a984-b24e-47f8-ae4e-7d8c9449fe6d" />

PR VK
<img width="640" height="256" alt="02978_f00001_fr1_00a00_C_32" src="https://github.com/user-attachments/assets/d0b737b0-b443-4188-b7d4-3ce54de4991f" />

"Hot Shots Golf FORE!_SCUS-97401_20221204040422.gs.xz" (see character portraits)

Master VK
<img width="640" height="224" alt="00017_f00001_fr1_00000_C_32" src="https://github.com/user-attachments/assets/226e9f33-c1ca-4331-84c1-c2b7dc5422df" />

PR VK
<img width="640" height="224" alt="00017_f00001_fr1_00000_C_32" src="https://github.com/user-attachments/assets/ba24cb7c-cc0a-4eaa-a303-5840ee5938ab" />

"Minna no Tennis_SCPS-15113_20230922202011.gs.xz"

Master VK
<img width="640" height="224" alt="00015_f00001_fr1_00000_C_32" src="https://github.com/user-attachments/assets/d1ba6dc5-8574-44b6-b670-43dddff29285" />

PR VK
<img width="640" height="224" alt="00015_f00001_fr1_00000_C_32" src="https://github.com/user-attachments/assets/efc4dda4-65a2-40c2-9f0a-c6ac0c183ab8" />

TJnotJT committed:
GS/HW: Depth feedback loops and accurate AFAIL.

TJnotJT commented:
Fixed barrier counts in VK and deleted some old commented code.

TJnotJT commented:
Latest push rebases and adds a temporary commit that for depth feedback via copy in OpenGL (based on the DX11 method). Based on a a few tests we have reason to believe that texture barriers for a depth attachment bound as a shader resource doesn't always work, so this is for testing until we can determine exactly.

TJnotJT commented:
Lastest commit rebases and adds a temporary commit to try a different depth feedback method for DX12 and OpenGL: convert Z to a temp color texture and attach it as a second RT (allows using recent barriers addition to DX12 (https://github.com/PCSX2/pcsx2/pull/13653) and barriers/FB fetch in OpenGL).

TJnotJT commented:
Removing draft status as dump runs on VK, DX12, GL, DX11 came clean with settings:
```
[EmuCore/GS]
...
DisableDepthFeedback = 0
accurate_blending_unit = 5
```
and VK with settings:
```
[EmuCore/GS]
...
DisableDepthFeedback = 1
accurate_blending_unit = 1
```
(The latter since there are refactors to alpha test that affect basic blend.)

lightningterror reviewed (changes requested):
Needs Metal changes as well.

lightningterror commented on pcsx2/GS/Renderers/DX11/GSDevice11.cpp:
```diff
@@ -2815,19 +2836,29 @@ void GSDevice11::RenderHW(GSHWDrawConfig& config)
+				CopyRect(draw_ds, draw_ds_clone, drawarea, drawarea.left, drawarea.top);
 			if (one_barrier || full_barrier)
-				PSSetShaderResource(2, draw_rt_clone);
+			{
```
Extra brackets aren't needed.

TellowKrinkle commented:
Some thoughts on the activation logic (note: the activation logic is complicated so I may not be understanding it correctly, let me know if it's not actually like this):
- It looks like you don't activate if dual source blend is in use and the renderer uses rt copies.  DSB is the less accurate blend option (vs software blend, which is more accurate).  Therefore you should be disabling DSB (and using SW blend) when feedback depth is in use.  In fact, you should probably *always* use SW blend when feedback depth is in use, since it's forcing barriers anyways.  This would mean doing a pre-check before EmulateBlending that decides whether Alpha Test is going to need barriers.
- You seem to treat feedback depth as free if something else is already requiring barriers.  Unlike with color, GPUs have a lot of depth-related optimizations (e.g. not running the shader at all if it knows the depth test will fail).  We might be already breaking some of them with our other shenanigans, but I doubt we break all of them on all GPUs all the time.  I'd avoid doing feedback depth if it doesn't improve accuracy, even if it seems free.

And for depth_feedback vs depth_as_rt_feedback in the device features, are they mutually exclusive?  If so, maybe make an enum for the options.  (I'll probably add a third option for the Metal fbfetch implementation(s).)

Finally, for depth_as_rt_feedback, you can probably save a resolve by binding and outputting to both depth and the feedback color binding in the draw.  Also did you do this in OGL because of the option for fbfetch or because OpenGL actually doesn't allow feedback writes to depth?  If the former, you might want to support both options in OpenGL, since many GPUs don't have fbfetch support and would be doing a bunch of extra copies for nothing.

Side note, these are my plans for Metal fbfetch on Apple GPUs in case any of its ideas are useful for your OGL implementation:
1. Bind the depth-as-rt depth texture as a memoryless render target, and the real depth texture as both a depth output and a shader resource.
2. Clear the depth-as-rt to -1 on render pass start, then in the shader, fbfetch read the depth-as-rt, and if it's -1, read the actual value from the real depth texture.  This should allow us to initialize the depth-as-rt texture without needing an initial copy.
3. Until the end of the render pass, for non-feedback draws, maintain the depth-as-rt texture by writing gl_FragCoord.z into it if zwrite is enabled.  This will require cooperation from GSRenderer to not use dual source blend while depth-as-rt is active.

(Not sure what I'll do for Intel fbfetch on Metal, might try the same as Apple GPUs or try doing a render copy into the texture like what you're currently doing in OGL.)

TJnotJT commented:
Thanks for the thorough feedback, my responses below.

> It looks like you don't activate if dual source blend is in use and the renderer uses rt copies. DSB is the less accurate blend option (vs software blend, which is more accurate). Therefore you should be disabling DSB (and using SW blend) when feedback depth is in use. In fact, you should probably always use SW blend when feedback depth is in use, since it's forcing barriers anyways. This would mean doing a pre-check before EmulateBlending that decides whether Alpha Test is going to need barriers.

Yes, currently, dual source blending will override using depth feedback with a RT copy. Though it isn't clear in the activation logic or in the PR description, I decided to make depth feedback dependent on the blend level (so dependent on whether SW blending is selected), since that's the closest thing we have to a performance slider for barrier usage/shader complexity. I can instead make depth feedback an independent option and have the blending selection depend on it as you've described.

> You seem to treat feedback depth as free if something else is already requiring barriers. Unlike with color, GPUs have a lot of depth-related optimizations (e.g. not running the shader at all if it knows the depth test will fail). We might be already breaking some of them with our other shenanigans, but I doubt we break all of them on all GPUs all the time. I'd avoid doing feedback depth if it doesn't improve accuracy, even if it seems free.

Good point, it probably shouldn't be considered free with barriers, especially for renderers that require a RT copy. Currently, if one of the accurate two pass methods (color-then-Z) are possible, we use those, and only take feedback as a replacement for the pass-then-fail two pass method. I guess this would go back to making depth feedback an independent option. Judging by dump runs, there are only a few games where depth feedback makes a visually noticable difference to accuracy, so it may be fine to make it an independent option and enable it just for games that need it.

> And for depth_feedback vs depth_as_rt_feedback in the device features, are they mutually exclusive? If so, maybe make an enum for the options. (I'll probably add a third option for the Metal fbfetch implementation(s).)

Yes, they are mutually exclusive so I can make them an enum.

> Finally, for depth_as_rt_feedback, you can probably save a resolve by binding and outputting to both depth and the feedback color binding in the draw. Also did you do this in OGL because of the option for fbfetch or because OpenGL actually doesn't allow feedback writes to depth? If the former, you might want to support both options in OpenGL, since many GPUs don't have fbfetch support and would be doing a bunch of extra copies for nothing.

For OGL, it was kind of a combination of both. Based on my system (Intel iGPU) and Jordan's, it seem like glTextureBarrier() did not properly synchronize depth textures. However, to be sure, I'll make a minimal test app to verify this in case something else was causing the issue with the dump we used. If we can determine that depth texture feedback is possible with OGL, we'll definitely that route.

Side note on DX12: we took the RT copy route after verifying with TheLastRar, who implemented aliasing barriers for DX12, that the spec didn't guarantee aliasing barriers for depth textures. However, I don't know if this may change with enhanced barriers.

> Side note, these are my plans for Metal fbfetch on Apple GPUs in case any of its ideas are useful for your OGL implementation: ...

This is a good idea and I'll probably revise OGL/DX12 to follow this outline. Even if they don't support simultaneous depth sample/write it could save the initial copy. For the actual mechanics of switching between depth/depth-as-rt, one possibility is to have each depth texture track of its own color copy and whether it's valid. I bring this up as I followed this approach for ROV, so any such framework could be shared, though of course you could do differently based on what is more efficent for the Metal implementation.

TJnotJT commented:
Converting to draft until revisions are complete.

TheLastRar commented:
> Side note on DX12: we took the RT copy route after verifying with TheLastRar, who implemented aliasing barriers for DX12, that the spec didn't guarantee aliasing barriers for depth textures. However, I don't know if this may change with enhanced barriers.

I'm afraid enhanced barriers doesn't change this from what I can tell.

TJnotJT commented:
> I'm afraid enhanced barriers doesn't change this from what I can tell.

I see, though either way it should improve the performance of both depth and color feedback.

Small update: It seems like OGL glTextureBarrier() works for depth buffers in a small test app on my system. I'm not sure what went wrong with earlier tests, but I'll try to reimplement this and see if the issue persists.

TJnotJT commented:
Updated with the following revisions:
- Make a new HWAccurateAlphaTest setting that is independent of blend setting. If alpha test requires feedback and uses a RT color copy it forces SW blend to avoid dual source blend conflict.
- Make the depth feedback functionality an enum (depth feedback for VK, GL, DX11; depth as color feedback for DX12).
- Consider feedback alpha test to be "free" if barriers are already enabled and only color feedback is needed for alpha test (compared to depth also previously).
- GL: Use depth feedback instead of depth as color feedback.
- DX12: In a depth feedback pass, output depth to both color copy and depth to avoid a resolve pass at the end.
- Add the accurate alpha test option to the UI in Settings>Graphics>Rendering.

Opening for review again as dumps runs were redone in DX12 with 
- maximum blending and HWAccurateAlphaTest=1
- basic blending and HWAccurateAlphaTest=0

Although GL was changed also, I'm unfortunately unable to fully test because depth barriers have sporadic issues on my sytem.

lightningterror commented:
So I tested Mesa llvmpipe software for Windows, Intel Haswell, GTX 1070, RX 570 and from the entire list only Intel Haswell was broken with DepthFeedbackMode = 2:

Depth feedback 2
<img width="1052" height="732" alt="gow gl intel depth 2" src="https://github.com/user-attachments/assets/7ebf2580-38ba-4c2f-ace4-97f5489fcd2d" />

Depth feedback 3 for comparison
<img width="1052" height="732" alt="gow gl intel depth 3" src="https://github.com/user-attachments/assets/250843eb-a0a1-4fb0-92b6-c7828ed153fe" />

lightningterror reviewed (commented):
Need to bump shader cache.

TellowKrinkle commented on pcsx2/GS/Renderers/HW/GSRendererHW.h:
```diff
@@ -273,4 +275,8 @@ class GSRendererHW : public GSRenderer
 	/// Compute the drawlist (if not already present) and bounding boxes for the current draw.
 	std::size_t ComputeDrawlistGetSize(float scale);
+
+	/// Create a temporary color clone of depth for depth feedback (DX12 only right now)
```
I assume this is no longer DX12 only?

TJnotJT replied:
Correct, I fixed the comment to include GL also.

TellowKrinkle commented on pcsx2/GS/Renderers/Common/GSDevice.h:
```diff
@@ -827,6 +885,13 @@ class GSDevice : public GSAlignedClass<32>
+	{
+		None,     // No support for depth feedback loops.
+		Depth,    // Implement depth feedback loops directly on a depth buffer.
+		DepthAsRT // Implement depth feedback loops by first converting depth to color RT.
```
```suggestion
		None,      // No support for depth feedback loops.
		Depth,     // Implement depth feedback loops directly on a depth buffer.
		DepthAsRT, // Implement depth feedback loops by first converting depth to color RT.
```

TJnotJT commented on pcsx2/GS/Renderers/HW/GSRendererHW.cpp:
```diff
@@ -8079,13 +8367,22 @@ __ri void GSRendererHW::DrawPrims(GSTextureCache::Target* rt, GSTextureCache::Ta
+		// but sometimes it slips through
 		if (m_conf.require_one_barrier || m_conf.require_full_barrier)
-			pxAssert(!m_conf.blend.enable);
+			pxAssert(!m_conf.blend.enable || m_conf.ps.no_color1);
```
I think as some point in development I was triggering this on OGL due to errors in the alpha test RGB_ONLY DSB method selection. You're right it shouldn't be needed, I reverted it back.

I'll retest GL to make sure that the path actually isn't taken.

TJnotJT commented:
Smoke tested VK, DX12, GL, DX11, and did a full dump run (logging only) on GL with FBfetch to make sure FBfetch + feedback + blending was never accidentally enabled simultaneously.

Rebased with minor merge conflicts.

TJnotJT commented:
Rebased and merged in DX11 changes from [#14132](https://github.com/PCSX2/pcsx2/pull/14132):
- Split out a separate flag for read-only DSV support (so it’s independent of the depth-sampling feature bit).
- Don’t use read-only DSV in depth feedback loops, since that’s effectively reading and writing depth at the same time.

Did a quick smoke test on DX11.

lightningterror reviewed (changes requested):
A better name would be read_only_hw_depth.

So there's an issue currently where Time Crisis 2, 3, Death by Degrees break when read only hw depth is not supported on dx11, on master we fallback to copies when it's not supported
https://github.com/PCSX2/pcsx2/blob/master/pcsx2/GS/Renderers/HW/GSRendererHW.cpp#L6890
and
https://github.com/PCSX2/pcsx2/blob/master/pcsx2/GS/Renderers/HW/GSRendererHW.cpp#L9680
Might be an issue for gl/vk too so we should use read_only_hw_depth  for these cases so that it can fallback to copies which is handled in RendererHW.

Alternatively we could ditch read_only_hw_depth , unbind dsv on all renderers and do depth testing in shader instead, no copies/barriers required, but maybe in a future pr.

TJnotJT committed:
Revert test_and_sample_depth to its original meaning of read-only DSV.

TJnotJT commented:
> A better name would be read_only_hw_depth.
> 
> So there's an issue currently where Time Crisis 2, 3, Death by Degrees break when read only hw depth is not supported on dx11, on master we fallback to copies when it's not supported https://github.com/PCSX2/pcsx2/blob/master/pcsx2/GS/Renderers/HW/GSRendererHW.cpp#L6890 and https://github.com/PCSX2/pcsx2/blob/master/pcsx2/GS/Renderers/HW/GSRendererHW.cpp#L9680 Might be an issue for gl/vk too so we should use read_only_hw_depth for these cases so that it can fallback to copies which is handled in RendererHW.
> 
> Alternatively we could ditch read_only_hw_depth , unbind dsv on all renderers and do depth testing in shader instead, no copies/barriers required, but maybe in a future pr.

It may be simpler to keep the original meaning of test_and_sample_depth to avoid extra changes in GSRendererHW. Let me know if the latest change looks good.

Shader depth testing usually requires at least one barrier (sometimes full barriers), similar to recursive blends.

TJnotJT commented:
Correct, DX12 only supports DepthFeedbackMode = 3.
